File: sdk/ts/src/addresses.ts
import { getAddress, isAddress, type Address } from 'viem';
import { NexusError } from './types.js';

/** Canonical ERC-8004 Identity Registry on Base mainnet. */
export const BASE_ERC8004_IDENTITY_REGISTRY: Address = '0x8004A169FB4a3325136EB29fA0ceB6D2e539a432';

/** Every contract address the SDK talks to. */
export interface Addresses {
  access: Address;
  identity: Address;
  reputation: Address;
  killSwitch: Address;
  auditLog: Address;
  feeRouter: Address;
  escrow: Address;
  paymentToken: Address;
}

/** Shape of `deployments/v2-<chainId>.json` as written by `script/v2/DeployCore.s.sol`. */
export interface DeploymentJson {
  chainId?: number;
  owner?: string;
  treasury?: string;
  paymentToken: string;
  AgentAccess: string;
  AgentIdentityV2: string;
  AgentReputationV2: string;
  AgentKillSwitchV2: string;
  AgentAuditLogV2: string;
  FeeRouter: string;
  AgentEscrowV2: string;
}

const FIELDS: ReadonlyArray<readonly [keyof Addresses, keyof DeploymentJson]> = [
  ['access', 'AgentAccess'],
  ['identity', 'AgentIdentityV2'],
  ['reputation', 'AgentReputationV2'],
  ['killSwitch', 'AgentKillSwitchV2'],
  ['auditLog', 'AgentAuditLogV2'],
  ['feeRouter', 'FeeRouter'],
  ['escrow', 'AgentEscrowV2'],
  ['paymentToken', 'paymentToken'],
];

function requireAddress(value: unknown, key: string): Address {
  if (typeof value !== 'string' || !isAddress(value)) {
    throw new NexusError(`Deployment JSON field "${key}" is not a valid address: ${String(value)}`);
  }
  return getAddress(value);
}

/**
 * Map a parsed `deployments/v2-*.json` object onto {@link Addresses}.
 * Throws {@link NexusError} when a field is missing or malformed.
 */
export function loadAddresses(json: unknown): Addresses {
  if (typeof json !== 'object' || json === null) {
    throw new NexusError('Deployment JSON must be an object');
  }
  const record = json as Record<string, unknown>;
  const partial: Partial<Record<keyof Addresses, Address>> = {};
  for (const [field, key] of FIELDS) {
    partial[field] = requireAddress(record[key], key);
  }
  return partial as Addresses;
}

/** Chain id recorded in a deployment JSON, when present. */
export function readChainId(json: unknown): number | undefined {
  if (typeof json !== 'object' || json === null) return undefined;
  const value = (json as Record<string, unknown>).chainId;
  return typeof value === 'number' ? value : undefined;
}

File: sdk/python/nexusweb3/addresses.py
"""Deployment addresses for a v2 core stack."""

from __future__ import annotations

import json
from dataclasses import dataclass, fields
from pathlib import Path
from typing import Any, Mapping

from eth_utils.address import to_checksum_address

from .types import ZERO_ADDRESS

#: Canonical ERC-8004 Identity Registry on Base. `AgentIdentityV2.linkERC8004` checks ownership here.
BASE_ERC8004_IDENTITY_REGISTRY = "0x8004A169FB4a3325136EB29fA0ceB6D2e539a432"

#: Maps `deployments/v2-*.json` keys onto :class:`Addresses` fields.
_JSON_KEYS: dict[str, str] = {
    "AgentAccess": "access",
    "AgentIdentityV2": "identity",
    "AgentReputationV2": "reputation",
    "AgentKillSwitchV2": "kill_switch",
    "AgentAuditLogV2": "audit_log",
    "FeeRouter": "fee_router",
    "AgentEscrowV2": "escrow",
    "paymentToken": "payment_token",
}

__all__ = ["Addresses", "load_addresses", "BASE_ERC8004_IDENTITY_REGISTRY"]


@dataclass(frozen=True)
class Addresses:
    """Checksummed addresses of one deployed v2 stack."""

    access: str
    identity: str
    reputation: str
    kill_switch: str
    audit_log: str
    fee_router: str
    escrow: str
    payment_token: str
    erc8004_registry: str = ZERO_ADDRESS

    def __post_init__(self) -> None:
        for field in fields(self):
            value = getattr(self, field.name)
            if not isinstance(value, str):
                raise TypeError(f"{field.name} must be a hex address string, got {type(value).__name__}")
            object.__setattr__(self, field.name, to_checksum_address(value))

    def to_dict(self) -> dict[str, str]:
        return {field.name: getattr(self, field.name) for field in fields(self)}


def _read_mapping(path_or_dict: str | Path | Mapping[str, Any]) -> Mapping[str, Any]:
    if isinstance(path_or_dict, Mapping):
        return path_or_dict
    path = Path(path_or_dict)
    if not path.is_file():
        raise FileNotFoundError(f"deployment file not found: {path}")
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise ValueError(f"{path} is not valid JSON: {exc}") from exc
    if not isinstance(data, Mapping):
        raise ValueError(f"{path} must contain a JSON object")
    return data


def load_addresses(path_or_dict: str | Path | Mapping[str, Any]) -> Addresses:
    """Load a `deployments/v2-<chain>.json` file (or an equivalent mapping) into :class:`Addresses`.

    Accepts both the deployment-file keys (``AgentAccess``, ``paymentToken``, …) and the
    dataclass field names (``access``, ``payment_token``, …).
    """
    data = _read_mapping(path_or_dict)
    kwargs: dict[str, str] = {}
    for json_key, field_name in _JSON_KEYS.items():
        if json_key in data:
            kwargs[field_name] = str(data[json_key])
        elif field_name in data:
            kwargs[field_name] = str(data[field_name])
        else:
            raise KeyError(f"deployment data is missing '{json_key}'")

    registry = data.get("erc8004Registry", data.get("erc8004_registry"))
    if registry:
        kwargs["erc8004_registry"] = str(registry)
    return Addresses(**kwargs)
